### PR TITLE
issue #367

### PR DESCRIPTION
**Pull request by lubennikovaav**
Next try.
Change CLI so that we always create node from scratch at 'pg start'.
This operation preserves previously existing config.

Add new flag '--config-only' to 'pg create'.
If this flag is passed, don't perform basebackup, just fill initial postgresql.conf for the node.

create_start() in test framework now does the following steps:
- pg create --config-only
- update config
- pg start - preserves config, calls basebackup

@LizardWizzard ,  @ololobus, could you do another pass of review, please?

### CODE REVIEW

**Review by knizhnik — APPROVED** [review]

**Review by ololobus — COMMENTED** [review]
I have no objections to do it this way in the zenith cli now, but let's decide whether we allow postgres restart or/and start after crash or not here https://github.com/zenithdb/zenith/issues/369

**Review by LizardWizzard — APPROVED** [review]

File: test_runner/fixtures/zenith_fixtures.py [hunk 1]
```diff
@@ -282,6 +282,14 @@ def start(self) -> 'Postgres':
         self.zenith_cli.run(['pg', 'start', self.branch, f'--tenantid={self.tenant_id}'])
         self.running = True
 
+        # # slurp port from postmaster.pid file in pgdata
```
Comment (LizardWizzard) [inline test_runner/fixtures/zenith_fixtures.py:285]: is this comment still needed?

File: control_plane/src/compute.rs [hunk 2]
```diff
@@ -407,7 +407,7 @@ impl PostgresNode {
 
     // Always destroy pgdatadir when we stop postgres
     pub fn stop(&self) -> Result<()> {
-        self.pg_ctl(&["stop"])?;
+        let _ = self.pg_ctl(&["stop"]);
```
Comment (LizardWizzard) [inline control_plane/src/compute.rs:410]: Is there a reason to ignore error here?

**Comment by lubennikovaav** [comment]
@ololobus , are you ok with this version?
Should we document this behavior somewhere? I haven't found a place better than code comments...

**Comment by ololobus** [comment]
> @ololobus , are you ok with this version?

I do not use current zenith cli in the cloud as it relies heavily on local distribution and directories structure. Anyway, current logic in this branch looks doubtful for me. It expects that postgres was shutdown 'cleanly', but you could never be sure that it's true. Postgres could be killed by OS, by user with `kill -9`, etc. So it doesn't look convenient. I would expect that `stop` just stops the cluster, but `start` ensures that it's a 'clean' start, i.e. if datadir exists then it's dropped and initialized again via baseback, if not then just initialized. In this case `restart` could be just an alias to `stop` + `start`.

> Should we document this behavior somewhere? I haven't found a place better than code comments...

Maybe we should start `docs/zenith_cli.md` and put this information there? But if it just works in any case, then these are internal implementation details and code comment is enough, IMO

**Comment by lubennikovaav** [comment]
> Anyway, current logic in this branch looks doubtful for me.

For me too.. Maybe we should align it with https://github.com/zenithdb/console/issues/119 
so that the console framework and test framework behave similarly?

>It expects that postgres was shutdown 'cleanly', but you could never be sure that it's true. 

I was thinking, that some external tool will be responsible for controlling such situations. I.e. if postgres failed, the console framework will stop/destroy VM and spawn a new one. But maybe CLI is the right place to perform database state checks.

> I would expect that `stop` just stops the cluster, but `start` ensures that it's a 'clean' start, i.e. if datadir exists then it's dropped and initialized again via baseback, if not then just initialized. In this case `restart` could be just an alias to `stop` + `start`.

It was like that in the previous iteration and didn't work with `zehith pg create`.
The test framework, for example, runs `zenith pg create` + apply config + `zenith pg start`. If the latter `start` command drops and reinitializes datadir, we just lose config changes and have no way to apply them.
I wonder if we should get rid of `zenith create` at all and instead just always pass config options to `pg start`? What do you think?

**Comment by ololobus** [comment]
> For me too.. Maybe we should align it with zenithdb/console#119
> so that the console framework and test framework behave similarly?

Purpose of zenithdb/console#119 is to deal with `systemd` and other fine-grained woes of a real installation. I do not think that we need to bring this complicated stuff into cli, at least for now.

> The test framework, for example, runs zenith pg create + apply config + zenith pg start. If the latter start command drops and
> reinitializes datadir, we just lose config changes and have no way to apply them.

But then simple `pg stop` + `pg start` should also drop the configs anyway, doesn't it?

If every start is intended to be a fresh start then what `zenith pg create` does currently beside creation of the datadir? Is the list of instances stored somewhere?

> I wonder if we should get rid of zenith create at all and instead just always pass config options to pg start? What do you think?

This should work, but maybe just teach `pg start` to copy pg_hba.conf and postgresql.conf before deletion of datadir if it exists?

**Comment by kelvich** [comment]
Another option is to override data directory on start with basebackup contents preserving postgresql.conf and friends untouched

File: control_plane/src/compute.rs [hunk 3]
```diff
@@ -279,20 +288,14 @@ impl PostgresNode {
...
-        //let mut ar = tar::Archive::new(copyreader);
-        let mut buf = vec![];
-        copyreader
-            .read_to_end(&mut buf)
-            .with_context(|| "reading base backup from page server failed")?;
-        let mut ar = tar::Archive::new(buf.as_slice());
-        ar.unpack(&pgdata)
-            .with_context(|| "extracting page backup failed")?;
+        if config_only {
+            //Just create an empty config file
```
Comment (ololobus) [inline control_plane/src/compute.rs:292]: I don't know why `cargo fmt` doesn't handle this, but I would add a space after `\\`

**Review by ololobus — COMMENTED** [review]
Logic around dropping compute on restart looks good to me now.

I have some doubts about `pg create --config-only`. What is it supposed to be used for?

File: control_plane/src/compute.rs [hunk 4]
```diff
@@ -85,48 +85,36 @@ impl ComputeControlPlane {
...
-            .pageserver
-            .branch_get_by_name(&tenantid, branch_name)?
-            .timeline_id;
-        let node = self.new_from_page_server(false, timeline_id, branch_name, tenantid)?;
         // Configure the node to stream WAL directly to the pageserver
         node.append_conf(
             "postgresql.conf",
             format!(
                 concat!(
-                    "shared_preload_libraries = zenith\n",
```
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply control_plane/src/compute.rs:129]: Yep, I noticed, that it is was added twice and removed an extra line.

**Comment by lubennikovaav** [comment]
> I have some doubts about `pg create --config-only`. What is it supposed to be used for?
It is used in `create_start()` function in test framework.

1 (old). `pg create` - performs basebackup and brings initial config 
1 (new) `pg create --config-only`  - only brings initial config 
2. apply config changes
3. `pg start` removes datadir (except config), performs basebackup, restores config

Without `--config-only` we would do unnecessary basebackup in step 1.

**Comment by ololobus** [comment]
> 1 (old). pg create - performs basebackup and brings initial config
> 1 (new) pg create --config-only - only brings initial config

So in most cases we actually need `pg create --config-only` now and `pg create` with basebackup is currently needed only for a couple of python tests. Maybe just reverse the `--config-only` flag meaning? For example, have `pg create` without basebackup by default (as with `--config-only` now) and `pg create --basebackup` (or `--with-basebackup`) for tests.

@alubennikova @kelvich @LizardWizzard any thoughts?

**Comment by kelvich** [comment]
> @alubennikova @kelvich @LizardWizzard any thoughts?

That is mostly for test, so any option works for me.

**Comment by lubennikovaav** [comment]
> @alubennikova @kelvich @LizardWizzard any thoughts?

All this CLI is for tests, so I wouldn't care that much.
I decided not to change the default behavior, as someone may rely on that.

**Pull request merged**